### ISSUE TITLE
Can decision tree based algorithms always find the best fit?

### ISSUE THREAD

**Issue by JoseAF** [google/yggdrasil-decision-forests#27]
Hi

I have a bit of a general question: when working with decision tree based algorithms such us Yggdrasil’s GBTs, should we expect that the algorithm will always find a better fit to the data (less error) when using all the available features as compared to when using a subset of them? The reasoning behind this assumption is that a good algorithm will always be able to choose the best features from within a group however big it is, so it’s reasonable to expect that if a subset of features gives the best fit, adding more features will not make the fit worse. It’s question about local versus global mínima. Any comments would be very appreciated…

**Comment by janpfeifer** [comment]
hi @JoseAF ,

Yes, that is the correct assumption. A good algorithm (and GBTs are one of them) will always (*) make good use of all features given, and automatically discard features that are not, and hence adding more features shouldn't make it worse **on the training data**. And that last part is important, because, in some cases, very good extra features on the training, turns out to lead to overfitting, and have worse results on *the test/validation data**. Generally (except in some degenerate cases) it may be the case that adding more features requires more regularization (fewer trees, trees less deep, larger minimum number of examples per leaf node, etc).

Also another thing to factor in during modeling: not all features come for free, so sometimes one may want to trim the features to make the model faster, or maybe avoid using some private data, etc.

cheers

(*) Another small exception here, that I'm guessing won't matter to you: most ML algorithms are a mix of greedy and random in nature: Decision Forests take greedily the best split, Neural Networks greedily go down the gradient of the loss, in a stochastic fashion (batch). I can imagine some really odd case where adding more features leads to a bad greedy interaction, but I would never expect this to happen in real life.

**Comment by JoseAF** [comment]
Hi @janpfeifer 

Thanks a lot for the quick reply. I'm slightly puzzled by a recent result I got using Yggdrasil's GBTs. I wonder if I may be hitting one of those unlikely cases, in real life!? I have a relatively simple image. If I train my model (default parameters and 10 trees) with 36 features I get both training and prediction errors that turn out to be higher than when I train the same model with a subset of 12 features taken out of those 36. I wondered whether the greedy nature of GBTs could explain this, but from your comment it sounds like you wouldn't expect this to happen in real life. I guess the other option that the model may be overfitting when using very good extra features doesn't apply here either, since in that case I would expect the training error to be much lower with those extra features, which is not the case. Any other comments/suggestions/ideas would be greatly appreciated :)

Thanks again.

**Comment by janpfeifer** [comment]
hi @JoseAF, that's odd ... I would have a few quick questions first to eliminate some simple cases:

* Is the difference small -- there can be some fluctuation due to the stochasticity, and hence the difference can be on that noise ?
* Can you add one feature at a time ? Is there one feature that causes the issue ?
* You mentioned the input is an image. Are these the values of pixels ? Are the 12 features being generated differently when you add the extra features ?

**Comment by JoseAF** [comment]
@janpfeifer Thanks for your comments. Yeah, I know it's odd... :( Regarding your questions:

1. I think the difference cannot be explained by stochastic fluctuation. I have observed fluctuations but in this case I run several simulations and the difference goes beyond anything I'd expect.

2. I have to investigate further which (if any one) feature is causing this issue. 

3. The values are pixels. The features are generated in the same way on both cases. 

Thanks again for your comments. I'll keep looking into this.

**Comment by janpfeifer** [comment]
I'm sorry @JoseAF, I'm stumped without looking at the data -- is it public data that you could make available ?

Maybe @achoum have some idea ?

**Comment by JoseAF** [comment]
@janpfeifer Unfortunately it's not public data... I'll keep looking into it and let you know if I find the issue. All your comments are much appreciated though. Thanks again for those

**Comment by Arnold1** [comment]
How do you measure overfitting?

**Comment by JoseAF** [comment]
A bug had creeped in which was messing up the results. All is sorted now. Thanks for the comments above!

**Comment by rstz** [comment]
Great to hear :) Closing this
